Adding initial version of the private_copies tutorial (#1)
justin-rosner/oneAPI-samples: DirectProgramming/DPC++FPGA/Tutorials/Features/private_copies/README.md

# Private Copies 
This FPGA tutorial explains how to use the private_copies attribute to improve loop concurrency.

***Documentation***:  The [DPC++ FPGA Code Samples Guide](https://software.intel.com/content/www/us/en/develop/articles/explore-dpcpp-through-intel-fpga-code-samples.html) helps you to navigate the samples and build your knowledge of DPC++ for FPGA. <br>
The [oneAPI DPC++ FPGA Optimization Guide](https://software.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide) is the reference manual for targeting FPGAs through DPC++. <br>
The [oneAPI Programming Guide](https://software.intel.com/en-us/oneapi-programming-guide) is a general resource for target-independent DPC++ programming.

| Optimized for                     | Description
---                                 |---

> | OS                                | Linux* Ubuntu* 18.04; Windows* 10

Reviewer: ```suggestion
| OS                                | Linux* Ubuntu* 18.04/20.04, RHEL*/CentOS* 8, SUSE* 15; Windows* 10

```

> | Hardware                          | Intel® Programmable Acceleration Card (PAC) with Intel Arria® 10 GX FPGA; <br> Intel® FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix® 10 SX)

Reviewer: ```suggestion
| Hardware                          | Intel® Programmable Acceleration Card (PAC) with Intel Arria® 10 GX FPGA <br> Intel® FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix® 10 SX) <br> Intel® FPGA 3rd party / custom platforms with oneAPI support <br> *__Note__: Intel® FPGA PAC hardware is only compatible with Ubuntu 18.04*

| Software                          | Intel® oneAPI DPC++ Compiler <br> Intel® FPGA Add-On for oneAPI Base Toolkit 

> | What you will learn               | The basic usage of the `private_copies` attribute <br> How the `private_copies` attribute affects loop throughput and resource use <br> How to apply the `private_copies` attribute to variables in your program <br> How to identify the correct `private_copies` factor for your program

Reviewer: ```suggestion
| What you will learn               | The basic usage of the `private_copies` attribute <br> How the `private_copies` attribute affects loop throughput and resource usage <br> How to apply the `private_copies` attribute to variables in your program <br> How to identify the correct `private_copies` factor for your program

```

| Time to complete                  | 15 minutes

## Purpose

> This tutorial demonstrates a simple example of applying the `private_copies` attribute to an array within a loop in a task kernel to trade off the on-chip memory use and throughput of the loop.

Reviewer: What information does "in a task kernel" brings ? I'm not sure to understand what you are implying

### Description of the `private_copies` Attribute

> The `private_copies` attribute enables you to control the number of copies of any memory that this is applied to. This works best when applying to any memory that has single iteration scope, thus enabling simultaneously executed loop iterations.

Reviewer: I'm not sure to understand what you mean by single iteration scope, 
Do you mean that the memory is instantiated in the most inner loop of a program?

#### Example: 

> Kernels in this tutorial design apply `[[intelfpga::private_copies(N)]]` to two arrays within the outer loop that are used by subsequent inner loops. These inner loops perform some simple math operations before storing the results. The following is an example of a loop nest:

Reviewer: So this would not fit in the best use case described above?

```
for (size_t i = 0; i < kMaxIter; i++) {
  [[intel::private_copies(2)]] int a[kSize];
  [[intel::private_copies(2)]] int b[kSize];
  int c[kSize];
  for (size_t j = 0; j < kSize; j++) {
    a[j] = accessor_array[(i * 4 + j) % kSize] * shift;
  }

>   for (size_t j = 0; j < kSize; j++)

Reviewer: I would suggest to use `{` and `}` for every loop

    b[j] = SimpleMath(a[j], a[kSize - j - 1]);
  for (size_t j = 0; j < kSize; j++)
    c[j] = SimpleMath(b[j], b[kSize - j - 1]);
  for (size_t j = 0; j < kSize; j++)
    r += c[j];
}  
```

> In this example, you only need to have two private copies of arrays a and b in order to have 2 concurrent outer loop iterations. The `private_copies` attribute in this example forces the compiler to create two private copies of the arrays `a` and `b`. Passing the parameter `N` to the `private_copies` attribute limits the number of privately-declared arrays that can be made in that loop. 

Reviewer: I find it non intuitive that you don't need 2 private copies of `c` as well
I'm guessing that this is because the compiler ends up implementing `c` as registersm but I can definitely see users getting confused about why `c` is treated differently
To fix this you could just replace `c` by an accessor write

> In this example, you only need to have two private copies of arrays a and b in order to have 2 concurrent outer loop iterations. The `private_copies` attribute in this example forces the compiler to create two private copies of the arrays `a` and `b`. Passing the parameter `N` to the `private_copies` attribute limits the number of privately-declared arrays that can be made in that loop. 

Reviewer: I think that this could benefit from a bit more detailed exaplanation (e,g, one copy in being read while the other is written to)

### Identifying the Correct `private_copies` Factor
Generally, increasing the number of private copies of an array within a loop situated in a task kernel will increase the throughput of that loop at the cost of increased memory usage. Additionally, in most cases there is a point at which increasing the number of private copies does not have any further effect on the throughput of the loop, as the maximum exploitable concurrency of that loop has been achieved. 

The correct `private_copies` factor for a given array depends on your goals for the design, the criticality of the loop in question, and its impact on your design's overall throughput. A typical design flow may be to: 
1. Experiment with different values of `private_copies`. 
2. Observe what impact the values have on the overall throughput and memory use of your design.

> 3. Choose the appropriate value that allows you to achieve your desired throughput and area goals.

Reviewer: I feel like this use model can be improved. Getting TP numbers takes a lot of time
Would there be a way for the compiler to tell the report what is the maximum private copies needed per memory?
Actually, I used `private_copies` before and just used a number that I new was too large, knowing that quartus would optimize it away.
Should the compiler know that more `private_copies` won't help, and just not implement them?

## Key Concepts
* The basic usage of the `private_copies` attribute 
* How the `private_copies` attribute affects loop throughput and resource use
* How to apply the `private_copies` attribute to variables in your program
* How to identify the correct `private_copies` factor for your program

## License  
Code samples are licensed under the MIT license. See
[License.txt](https://github.com/oneapi-src/oneAPI-samples/blob/master/License.txt) for details.

Third party program Licenses can be found here: [third-party-programs.txt](https://github.com/oneapi-src/oneAPI-samples/blob/master/third-party-programs.txt)

## Building the `private_copies` Tutorial

> 

Reviewer: ```suggestion

> **Note**: If you have not already done so, set up your CLI
> environment by sourcing  the `setvars` script located in
> the root of your oneAPI installation.
>
> Linux Sudo: . /opt/intel/oneapi/setvars.sh
>
> Linux User: . ~/intel/oneapi/setvars.sh
>
> Windows: C:\Program Files(x86)\Intel\oneAPI\setvars.bat
>
>For more information on environment variables, see Use the setvars Script for [Linux or macOS](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-linux-or-macos.html), or [Windows](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-windows.html).
```

### Include Files
The included header `dpc_common.hpp` is located at `%ONEAPI_ROOT%\dev-utilities\latest\include` on your development system.

### Running Samples in DevCloud
If running a sample in the Intel DevCloud, remember that you must specify the compute node (fpga_compile, fpga_runtime:arria10, or fpga_runtime:stratix10) and run in batch or interactive mode. For more information, see the Intel® oneAPI Base Toolkit Get Started Guide ([https://devcloud.intel.com/oneapi/documentation/base-toolkit/](https://devcloud.intel.com/oneapi/documentation/base-toolkit/)).

When compiling for FPGA hardware, it is recommended to increase the job timeout to 12h.

> 

Reviewer: ```suggestion

### Using Visual Studio Code*  (Optional)

You can use Visual Studio Code (VS Code) extensions to set your environment, create launch configurations,
and browse and download samples.

The basic steps to build and run a sample using VS Code include:
 - Download a sample using the extension **Code Sample Browser for Intel oneAPI Toolkits**.
 - Configure the oneAPI environment with the extension **Environment Configurator for Intel oneAPI Toolkits**.
 - Open a Terminal in VS Code (**Terminal>New Terminal**).
 - Run the sample in the VS Code terminal using the instructions below.

To learn more about the extensions and how to configure the oneAPI environment, see
[Using Visual Studio Code with Intel® oneAPI Toolkits](https://software.intel.com/content/www/us/en/develop/documentation/using-vs-code-with-intel-oneapi/top.html).

After learning how to use the extensions for Intel oneAPI Toolkits, return to this readme for instructions on how to build and run a sample.
```

### On a Linux* System

1. Generate the `Makefile` by running `cmake`.
     ```
   mkdir build
   cd build
   ```
   To compile for the Intel® PAC with Intel Arria® 10 GX FPGA, run `cmake` using the command:  
    ```
    cmake ..
   ```
   Alternatively, to compile for the Intel® FPGA PAC D5005 (with Intel Stratix® 10 SX), run `cmake` using the command:

   ```
   cmake .. -DFPGA_BOARD=intel_s10sx_pac:pac_s10
   ```

> 

Reviewer: ```suggestion

   You can also compile for a custom FPGA platform. Ensure that the board support package is installed on your system. Then run `cmake` using the command:
   
   cmake .. -DFPGA_BOARD=<board-support-package>:<board-variant>
   
```
You need to add the ` ``` `

2. Compile the design through the generated `Makefile`. The following build targets are provided, matching the recommended development flow:

   * Compile for emulation (fast compile time, targets emulated FPGA device): 
      ```
      make fpga_emu
      ```
   * Generate the optimization report: 
     ```
     make report
     ``` 
   * Compile for FPGA hardware (longer compile time, targets FPGA device): 
     ```
     make fpga
     ``` 
3. (Optional) As the FPGA hardware compile may take several hours to complete, FPGA precompiled binaries (compatible with Linux* Ubuntu* 18.04) can be downloaded <a href="https://iotdk.intel.com/fpga-precompiled-binaries/latest/private_copies.fpga.tar.gz" download>here</a>.

### On a Windows* System

1. Generate the `Makefile` by running `cmake`.
     ```
   mkdir build
   cd build
   ```
   To compile for the Intel® PAC with Intel Arria® 10 GX FPGA, run `cmake` using the command:  
    ```
    cmake -G "NMake Makefiles" ..
   ```
   Alternatively, to compile for the Intel® FPGA PAC D5005 (with Intel Stratix® 10 SX), run `cmake` using the command:

   ```
   cmake -G "NMake Makefiles" .. -DFPGA_BOARD=intel_s10sx_pac:pac_s10
   ```

> 

Reviewer: ```suggestion

   You can also compile for a custom FPGA platform. Ensure that the board support package is installed on your system. Then run `cmake` using the command:
   
   cmake -G "NMake Makefiles" .. -DFPGA_BOARD=<board-support-package>:<board-variant>
   
```
You need to add the ` ``` `

2. Compile the design through the generated `Makefile`. The following build targets are provided, matching the recommended development flow:

   * Compile for emulation (fast compile time, targets emulated FPGA device): 
     ```
     nmake fpga_emu
     ```
   * Generate the optimization report: 
     ```
     nmake report
     ``` 

>    * An FPGA hardware target is not provided on Windows*. 

Reviewer: ```suggestion
      * Compile for FPGA hardware (longer compile time, targets FPGA device):

     nmake fpga

```

You need to add the ` ``` `

> *Note:* The Intel® PAC with Intel Arria® 10 GX FPGA and Intel® FPGA PAC D5005 (with Intel Stratix® 10 SX) do not yet support Windows*. Compiling to FPGA hardware on Windows* requires a third-party or custom Board Support Package (BSP) with Windows* support.

Reviewer: ```suggestion
*Note:* The Intel® PAC with Intel Arria® 10 GX FPGA and Intel® FPGA PAC D5005
(with Intel Stratix® 10 SX) do not support Windows*. Compiling to FPGA hardware
on Windows* requires a third-party or custom Board Support Package (BSP) with
Windows* support.<br> *Note:* If you encounter any issues with long paths when
compiling under Windows*, you may have to create your ‘build’ directory in a
shorter path, for example c:\samples\build. You can then run cmake from that
directory, and provide cmake with the full path to your sample directory.

```

>  

Reviewer: ```suggestion
 
### Troubleshooting
If an error occurs, you can get more details by running `make` with
the `VERBOSE=1` argument:
``make VERBOSE=1``
For more comprehensive troubleshooting, use the Diagnostics Utility for
Intel® oneAPI Toolkits, which provides system checks to find missing
dependencies and permissions errors.
[Learn more](https://software.intel.com/content/www/us/en/develop/documentation/diagnostic-utility-user-guide/top.html).
```

 ### In Third-Party Integrated Development Environments (IDEs)

> You can compile and run this tutorial in the Eclipse* IDE (in Linux*) and the Visual Studio* IDE (in Windows*). For instructions, refer to the following link: [Intel® oneAPI DPC++ FPGA Workflows on Third-Party IDEs](https://software.intel.com/en-us/articles/intel-oneapi-dpcpp-fpga-workflow-on-ide)

Reviewer: ```suggestion
You can compile and run this tutorial in the Eclipse* IDE (in Linux*) and the Visual Studio* IDE (in Windows*). For instructions, refer to the following link: [Intel® oneAPI DPC++ FPGA Workflows on Third-Party IDEs](https://www.intel.com/content/www/us/en/developer/articles/technical/intel-oneapi-dpcpp-fpga-workflow-on-ide.html)

```

## Examining the Reports
Locate `report.html` in the `private_copies_report.prj/reports/` or `private_copies_s10_pac_report.prj/reports/` directory. Open the report in any of Chrome*, Firefox*, Edge*, or Internet Explorer*.

> On the main report page, scroll down to the section titled "Estimated Resource Usage". Each kernel name ends in the private_copies attribute argument used for that kernel, e.g., `kernelCompute1` uses a private_copies attribute value of 1. You can verify that the number of RAMs used for each kernel increases with the private_copies value used, with the exception of private_copies 0, which instructs the compiler to choose a default value.

Reviewer: What is the default value? Does the compiler try to find a sensible value? Does it try to optimize TP?

## Running the Sample

 1. Run the sample on the FPGA emulator (the kernel executes on the CPU):
     ```
     ./private_copies.fpga_emu     (Linux)
     private_copies.fpga_emu.exe   (Windows)
     ```
2. Run the sample on the FPGA device:
     ```

>      ./private_copies.fpga         (Linux)

Reviewer: ```suggestion
     ./private_copies.fpga         (Linux)
     private_copies.fpga.exe     (Windows)

```
